File: run_bench/saturate.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <time.h>

#define N (1 << 28) // 256 MB
#define ITERATIONS 100

int main(void) {
    // Allocate memory
    uint8_t *data = (uint8_t *) malloc(N);
    if (data == NULL) {
        fprintf(stderr, "Failed to allocate memory\n");
        return 1;
    }

    // Initialize memory
    for (size_t i = 0; i < N; i++) {
        data[i] = (uint8_t) (i % 256);
    }

    // Measure memory bandwidth
    double start_time = (double) clock() / CLOCKS_PER_SEC;
    for (int i = 0; i < ITERATIONS; i++) {
        for (size_t j = 0; j < N; j++) {
            data[j]++;
        }
    }
    double end_time = (double) clock() / CLOCKS_PER_SEC;

    // Compute memory bandwidth
    double elapsed_time = end_time - start_time;
    size_t bytes_processed = ITERATIONS * N;
    double bandwidth = bytes_processed / elapsed_time / (1 << 30); // GB/s
    printf("Memory bandwidth: %.2f GB/s\n", bandwidth);

    // Free memory
    free(data);

    return 0;
}

File: run_bench/saturate.py
import time
import numpy as np
import threading

# Allocate two large arrays of floats
num_elements = 10_000_000  # 10 million floats
data1 = np.ones(num_elements, dtype=np.float32)
data2 = np.ones(num_elements, dtype=np.float32)

# Define a function that reads and writes the entire array
num_iterations = 1000
def memory_bandwidth_test(data):
    for i in range(num_iterations):
        # Read the entire array
        total = np.sum(data)

        # Write to the entire array
        data += 1.0

# Create two threads that run the memory_bandwidth_test() function
threads_list = []
for t in range(2):
    thread = threading.Thread(target=memory_bandwidth_test, args=(np.ones(num_elements, dtype=np.float32),))
    threads_list.append(thread)
    # thread2 = threading.Thread(target=memory_bandwidth_test, args=(data2,))

# Start the threads and measure the elapsed time
start_time = time.monotonic()
for each_thread in threads_list:
    each_thread.start()
# thread1.start()
# thread2.start()
for each_thread in threads_list:
    each_thread.join()
# thread1.join()
# thread2.join()
end_time = time.monotonic()
elapsed_time = end_time - start_time

# Compute the memory bandwidth in GB/s
bytes_processed = 2 * num_iterations * num_elements * 4  # 4 bytes per float
bandwidth = bytes_processed / elapsed_time / 1e9
print(f"Memory bandwidth: {bandwidth:.2f} GB/s")
